### DISCUSSION TITLE
Suggestions for how to get trepan3k speaking/working with DAP

### DISCUSSION THREAD

**Discussion by rocky** [microsoft/debugpy#1731] [Q&A]
I would like to investigate getting the [Trepan debuggers](https://github.com/Trepan-Debuggers/) , and more specifically [trepan3k](https://pypi.org/project/trepan3k/) speaking DAP. 

I would appreciate suggestions for learning how to write a DAP adapter for trepan3k and how to approach getting this done in an easy,  incremental path. (Of course, I've started looking at the Wiki) 

Looking at the code, I see lots of sophisticated stuff, like the language for testing which accommodates variations that happen normally in responses. 

Thanks.

_I might be in the wrong place here since this seems to be client code and I want to create a DAP-enabled Python server attached to trepan3k?_

**Comment by rchiodo** [comment]
The code in the debugpy.adapter is the part that speaks DAP with VS code. For example, this is where we handle the initialize request:

https://github.com/microsoft/debugpy/blob/3573ca0e2bb562e8a0d3de68e531c3419a06c82a/src/debugpy/adapter/clients.py#L148.

Handling DAP is basically just listening to the messages and translating that into different things in the debugger. For example, this is where a thread suspend event is sent (like when hitting a breakpoint):

https://github.com/microsoft/debugpy/blob/3573ca0e2bb562e8a0d3de68e531c3419a06c82a/src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_net_command_factory_json.py#L389

Debugpy would then forward that message onto VS code.
